File: packages/core/cleanup.js
/**
 * Matrix REST API cleanup utilities.
 * Uses fetch() directly — works in both Node.js 22+ and browser.
 */

/**
 * Parse an --older-than duration string into a cutoff timestamp (ms).
 * Supports: Nd (days), Nw (weeks), Nm (months = 30 days).
 * Returns null if input is falsy.
 * @param {string} [str] - e.g. "2w", "1m", "7d"
 * @returns {number|null} Unix timestamp cutoff in ms, or null
 */
export function parseOlderThan(str) {
  if (!str) return null;
  const match = str.trim().match(/^(\d+)([dwm])$/i);
  if (!match) throw new Error(`Invalid --older-than format: "${str}". Use Nd, Nw, or Nm (e.g. 7d, 2w, 1m).`);
  const n = parseInt(match[1], 10);
  const unit = match[2].toLowerCase();
  const msPerDay = 86400000;
  const multipliers = { d: msPerDay, w: 7 * msPerDay, m: 30 * msPerDay };
  return Date.now() - (n * multipliers[unit]);
}

async function matrixFetch(homeserverUrl, path, accessToken, options = {}, _retries = 0) {
  const url = `${homeserverUrl}${path}`;
  const resp = await fetch(url, {
    ...options,
    headers: {
      'Authorization': `Bearer ${accessToken}`,
      'Content-Type': 'application/json',
      ...options.headers,
    },
  });
  if (resp.status === 429 && _retries < 3) {
    const retryAfter = parseInt(resp.headers.get('Retry-After') || '5', 10);
    await new Promise(r => setTimeout(r, retryAfter * 1000));
    return matrixFetch(homeserverUrl, path, accessToken, options, _retries + 1);
  }
  return resp;
}

/**
 * Clean up old devices. Deletes all devices except the current one.
 * Requires password for UIA (User-Interactive Authentication).
 *
 * @param {object} opts
 * @param {string} opts.accessToken
 * @param {string} opts.homeserverUrl
 * @param {string} opts.currentDeviceId - Will NOT be deleted
 * @param {string} opts.userId - Full Matrix user ID (e.g. @user:matrix.org)
 * @param {string} opts.password - Required for UIA
 * @param {number} [opts.olderThan] - Cutoff timestamp (ms). Only delete devices last seen before this. From parseOlderThan().
 * @param {function} [opts.onProgress] - Called with status messages
 * @returns {{ preview: Array<{device_id, display_name, last_seen_ts}>, execute: function }}
 */
export async function cleanupDevices({
  accessToken, homeserverUrl, currentDeviceId, userId, password, olderThan, onProgress = () => {},
}) {
  onProgress('Fetching device list...');
  const resp = await matrixFetch(homeserverUrl, '/_matrix/client/v3/devices', accessToken);
  if (!resp.ok) throw new Error(`Failed to list devices: ${resp.status}`);
  const { devices } = await resp.json();

  let toDelete = devices.filter(d => d.device_id !== currentDeviceId);
  if (olderThan) {
    toDelete = toDelete.filter(d => d.last_seen_ts && d.last_seen_ts < olderThan);
  }

  return {
    preview: toDelete.map(d => ({
      device_id: d.device_id,
      display_name: d.display_name || '(unnamed)',
      last_seen_ts: d.last_seen_ts,
      last_seen_ip: d.last_seen_ip,
    })),
    execute: async () => {
      if (toDelete.length === 0) {
        onProgress('No devices to delete');
        return { deleted: 0 };
      }

      const deviceIds = toDelete.map(d => d.device_id);
      onProgress(`Deleting ${deviceIds.length} device(s)...`);

      // First request without auth — gets UIA session
      const resp1 = await matrixFetch(
        homeserverUrl, '/_matrix/client/v3/delete_devices', accessToken,
        { method: 'POST', body: JSON.stringify({ devices: deviceIds }) },
      );

      if (resp1.ok) {
        onProgress(`Deleted ${deviceIds.length} device(s)`);
        return { deleted: deviceIds.length };
      }

      if (resp1.status !== 401) {
        throw new Error(`Device deletion failed: ${resp1.status} ${await resp1.text()}`);
      }

      // UIA required — extract session and retry with password
      const uiaInfo = await resp1.json();
      const session = uiaInfo.session;

      const localpart = userId.split(':')[0].replace('@', '');
      const resp2 = await matrixFetch(
        homeserverUrl, '/_matrix/client/v3/delete_devices', accessToken,
        {
          method: 'POST',
          body: JSON.stringify({
            devices: deviceIds,
            auth: {
              type: 'm.login.password',
              identifier: { type: 'm.id.user', user: localpart },
              password,
              session,
            },
          }),
        },
      );

      if (!resp2.ok) {
        const errBody = await resp2.text();
        throw new Error(`Device deletion UIA failed: ${resp2.status} ${errBody}`);
      }

      onProgress(`Deleted ${deviceIds.length} device(s)`);
      return { deleted: deviceIds.length };
    },
  };
}

File: crates/mxdx-client/src/cleanup.rs
use anyhow::Result;
use serde::Deserialize;

#[derive(Deserialize)]
struct DeviceInfo {
    device_id: String,
    display_name: Option<String>,
    last_seen_ts: Option<u64>,
}

#[derive(Deserialize)]
struct DevicesResponse {
    devices: Vec<DeviceInfo>,
}

/// List all devices for the current user.
async fn list_devices(homeserver: &str, access_token: &str) -> Result<Vec<DeviceInfo>> {
    let client = reqwest::Client::new();
    let url = format!(
        "{}/_matrix/client/v3/devices",
        homeserver.trim_end_matches('/')
    );
    let resp = client.get(&url).bearer_auth(access_token).send().await?;
    if !resp.status().is_success() {
        anyhow::bail!("Failed to list devices: {}", resp.status());
    }
    let data: DevicesResponse = resp.json().await?;
    Ok(data.devices)
}

/// Delete a device with UIA password authentication.
async fn delete_device(
    homeserver: &str,
    access_token: &str,
    device_id: &str,
    user_id: &str,
    password: &str,
) -> Result<()> {
    let client = reqwest::Client::new();
    let base = homeserver.trim_end_matches('/');
    let url = format!("{}/_matrix/client/v3/devices/{}", base, device_id);

    // First attempt without auth — may succeed or return 401 with UIA session
    let resp = client.delete(&url).bearer_auth(access_token).send().await?;
    if resp.status().is_success() {
        return Ok(());
    }
    if resp.status().as_u16() != 401 {
        anyhow::bail!("Delete device failed: {}", resp.status());
    }

    // UIA required — extract session
    let uia_info: serde_json::Value = resp.json().await?;
    let session = uia_info["session"].as_str().unwrap_or("");
    let localpart = user_id
        .split(':')
        .next()
        .unwrap_or(user_id)
        .trim_start_matches('@');

    let auth_body = serde_json::json!({
        "auth": {
            "type": "m.login.password",
            "identifier": { "type": "m.id.user", "user": localpart },
            "password": password,
            "session": session,
        }
    });

    let resp2 = client
        .delete(&url)
        .bearer_auth(access_token)
        .json(&auth_body)
        .send()
        .await?;
    if !resp2.status().is_success() {
        anyhow::bail!("UIA delete failed: {}", resp2.status());
    }
    Ok(())
}

/// List joined rooms.
async fn list_joined_rooms(homeserver: &str, access_token: &str) -> Result<Vec<String>> {
    let client = reqwest::Client::new();
    let url = format!(
        "{}/_matrix/client/v3/joined_rooms",
        homeserver.trim_end_matches('/')
    );
    let resp = client.get(&url).bearer_auth(access_token).send().await?;
    if !resp.status().is_success() {
        anyhow::bail!("Failed to list rooms: {}", resp.status());
    }
    let data: serde_json::Value = resp.json().await?;
    let rooms = data["joined_rooms"]
        .as_array()
        .map(|arr| {
            arr.iter()
                .filter_map(|v| v.as_str().map(String::from))
                .collect()
        })
        .unwrap_or_default();
    Ok(rooms)
}

/// Leave and forget a room.
async fn leave_and_forget(homeserver: &str, access_token: &str, room_id: &str) -> Result<()> {
    let client = reqwest::Client::new();
    let base = homeserver.trim_end_matches('/');
    let encoded = urlencoding::encode(room_id);

    let leave_url = format!("{}/_matrix/client/v3/rooms/{}/leave", base, encoded);
    let _ = client
        .post(&leave_url)
        .bearer_auth(access_token)
        .json(&serde_json::json!({}))
        .send()
        .await?;

    let forget_url = format!("{}/_matrix/client/v3/rooms/{}/forget", base, encoded);
    let _ = client
        .post(&forget_url)
        .bearer_auth(access_token)
        .json(&serde_json::json!({}))
        .send()
        .await?;

    Ok(())
}

/// Logout all sessions (nuclear — deletes all devices).
async fn logout_all(homeserver: &str, access_token: &str) -> Result<()> {
    let client = reqwest::Client::new();
    let url = format!(
        "{}/_matrix/client/v3/logout/all",
        homeserver.trim_end_matches('/')
    );
    let resp = client
        .post(&url)
        .bearer_auth(access_token)
        .json(&serde_json::json!({}))
        .send()
        .await?;
    if !resp.status().is_success() {
        let body = resp.text().await.unwrap_or_default();
        anyhow::bail!("logout/all failed: {}", body);
    }
    Ok(())
}

/// Run the cleanup command.
pub async fn run_cleanup(
    homeserver: &str,
    access_token: &str,
    current_device_id: &str,
    user_id: &str,
    password: &str,
    targets: &str,
    force: bool,
    delete_all_sessions: bool,
) -> Result<()> {
    let target_list: Vec<&str> = targets.split(',').map(|s| s.trim()).collect();

    if delete_all_sessions {
        if !force {
            eprintln!("WARNING: This will log out ALL sessions and delete ALL devices.");
            eprintln!("You will need to re-login. Use --force to skip this prompt.");
            return Ok(());
        }
        eprintln!("Logging out all sessions...");
        logout_all(homeserver, access_token).await?;
        eprintln!("All sessions logged out. All devices deleted. Re-login required.");
        return Ok(());
    }

    let do_devices = target_list.iter().any(|t| *t == "devices" || *t == "all");
    let do_rooms = target_list.iter().any(|t| *t == "rooms" || *t == "all");

    // Validate targets
    for target in &target_list {
        match *target {
            "devices" | "rooms" | "all" => {}
            other => {
                eprintln!(
                    "Unknown cleanup target: '{}'. Use: devices, rooms, all",
                    other
                );
            }
        }
    }

    if do_devices {
        eprintln!("Cleaning up devices...");
        let devices = list_devices(homeserver, access_token).await?;
        let to_delete: Vec<&DeviceInfo> = devices
            .iter()
            .filter(|d| d.device_id != current_device_id)
            .collect();

        eprintln!(
            "Found {} device(s) to delete (keeping current: {})",
            to_delete.len(),
            current_device_id
        );
        for d in &to_delete {
            let name = d.display_name.as_deref().unwrap_or("(unnamed)");
            let ts = d
                .last_seen_ts
                .map(|t| format!("{}ms", t))
                .unwrap_or_else(|| "never".into());
            eprintln!("  {} — {} (last seen: {})", d.device_id, name, ts);
        }

        if !force && !to_delete.is_empty() {
            eprintln!("Use --force to proceed with deletion.");
        } else {
            let mut deleted = 0;
            let mut errors = 0;
            for d in &to_delete {
                match delete_device(homeserver, access_token, &d.device_id, user_id, password).await
                {
                    Ok(()) => {
                        deleted += 1;
                    }
                    Err(e) => {
                        eprintln!("  Error deleting {}: {}", d.device_id, e);
                        errors += 1;
                    }
                }
            }
            eprintln!("Devices: {} deleted, {} errors", deleted, errors);
        }
    }

    if do_rooms {
        eprintln!("Cleaning up rooms...");
        let rooms = list_joined_rooms(homeserver, access_token).await?;
        eprintln!("Found {} joined room(s)", rooms.len());

        if !force && !rooms.is_empty() {
            for r in &rooms {
                eprintln!("  {}", r);
            }
            eprintln!("Use --force to leave and forget all rooms.");
        } else {
            let mut left = 0;
            let mut errors = 0;
            for room_id in &rooms {
                match leave_and_forget(homeserver, access_token, room_id).await {
                    Ok(()) => {
                        eprintln!("  Left {}", room_id);
                        left += 1;
                    }
                    Err(e) => {
                        eprintln!("  Error on {}: {}", room_id, e);
                        errors += 1;
                    }
                }
            }
            eprintln!("Rooms: {} left+forgotten, {} errors", left, errors);
        }
    }

    Ok(())
}
